tex/gregoriotex-signs.tex

```diff
@@ -2695,16 +2697,59 @@
 
 % the top level macro:
 % #1 is the height
-% #2 is the char of the alteration
-% #3 is the char of the alteration hole
-% #4 is 1 if the alteration is the flat in a clef
-% #5 are the signs to typeset before the glyph (typically additional bars, as they must be "behind" the glyph)
-% #6 are the signs to typeset after the glyph (almost all signs)
-% #7 is the line:char:column for a textedit link
-\def\gre@alteration#1#2#3#4#5#6#7{%
-  \gre@trace{gre@alteration{#1}{#2}{#3}{#4}{#5}{#6}{#7}}%
-  \setbox\gre@box@temp@width=\hbox{\gre@pointandclick{#2}{#7}}%
-  \ifnum#4=0\relax %
+% #2 is the alteration type: 0 = none, 1 = flat, 2 = natural, 3 = sharp,
+%                            4 = parenthesized flat, 5 = parenthesized natural,
+%                            6 = parenthesized flat
+% #3 is 1 if the alteration is the flat in a clef
+% #4 are the signs to typeset before the glyph (typically additional bars, as they must be "behind" the glyph)
+% #5 are the signs to typeset after the glyph (almost all signs)
+% #6 is the line:char:column for a textedit link
+\def\gre@alteration#1#2#3#4#5#6{%
+  \gre@trace{gre@alteration{#1}{#2}{#3}{#4}{#5}{#6}}%
+  \global\advance\gre@attr@alteration@id by 1\relax
+  \gre@alteration@isvisible{0}%
+  \ifgre@alteration@isvisible
+    \gre@alteration@visible{#1}{#2}{#3}{#4}{#5}{#6}%
+  \else
+    % ignore #4 (ledger line) and #5 (not used anyway?)
+    \hbox
+      attr \gre@attrid@alteration@height=#1\relax
+      attr \gre@attrid@alteration@type=#2\relax
+      {}%
+  \fi
+}
+
+% Test whether an alteration should be visible.
+% #1 = 0 for current alteration, 1 for next alteration.
+% Sets \ifgre@alteration@isvisibletrue or \ifgre@alteration@isvisiblefalse.
+\newif\ifgre@alteration@isvisible
+\def\gre@alteration@isvisible#1{%
+  \ifgre@alteration@line
+    \ifcase\directlua{gregoriotex.is_first_alteration(#1)}\relax
+      \global\gre@alteration@isvisibletrue
+    \else
+      \global\gre@alteration@isvisiblefalse
+    \fi
+  \else
+    \global\gre@alteration@isvisibletrue
+  \fi
+}
+
+\def\gre@alteration@visible#1#2#3#4#5#6{%
+  \gre@trace{gre@alteration@visible{#1}{#2}{#3}{#4}{#5}{#6}}%
+  \ifcase#2\relax
+    \or\def\alterationchar{\gre@fontchar@flat}\def\alterationhole{\gre@fontchar@flathole}% flat
```

Comment:
We try to avoid unprefixed macros in order to prevent name conflicts with other packages (regardless of how unlikely we think they are).  Our prefix system is:
- `\gre@`: Internal macros that users should never interact with in their documents
- `\Gre`: Macros written to the gtex file by the executable. Advanced users might interact with these to do special things in their document
- `\gre`: user facing macros.  We expect users to make use of these macros regularly.

With that in mind, `\alterationchar` and `\alterationhole` should probably be named `\gre@alterationchar` and `\gre@alterationhole`.